src/Shared/Buffers/RefPooledArrayBufferWriter.cs

```diff
@@ -0,0 +1,266 @@
+    /// </summary>
+    /// <param name="initialBuffer">The initial buffer to start writer with.</param>
+    public RefPooledArrayBufferWriter(Span<T> initialBuffer)
+    {
+        // no rented buffer initially - only if we need to grow over the limits of initialBuffer
+        _rentedBuffer = null!;
+
+        _buffer = initialBuffer;
+        _index = 0;
+    }
+
+    /// <summary>
+    /// Clears the buffer contents and returns the rented array to the ArrayPool.
+    /// This must be called to properly clean up resources.
+    /// </summary>
+    public void Dispose()
+    {
+        // to avoid `bool isDisposed` field, we can use negative index as disposed marker
+        _index = -1;
+
+        if (_rentedBuffer is not null)
+        {
+            // Optionally clear the buffer before returning to pool (security)
+            // Uncomment if needed for sensitive data:
+            // _buffer.AsSpan(0, _index).Clear();
+
+            ArrayPool<T>.Shared.Return(_rentedBuffer, clearArray: false);
+            _buffer = null!;
+        }
+    }
+
+    /// <summary>
+    /// Gets the number of bytes written to the buffer.
+    /// </summary>
+    public int WrittenCount
+    {
+        get
+        {
+            ThrowIfDisposed();
+            return _index;
+        }
+    }
+
+    /// <summary>
+    /// Gets the capacity of the underlying buffer.
+    /// </summary>
+    public int Capacity
+    {
+        get
+        {
+            ThrowIfDisposed();
+            return _rentedBuffer is not null ? _rentedBuffer.Length : _buffer.Length;
+        }
+    }
+
+    /// <summary>
+    /// Gets the available space remaining in the buffer.
+    /// </summary>
+    public int FreeCapacity
+    {
+        get
+        {
+            ThrowIfDisposed();
+            var length = _rentedBuffer is not null ? _rentedBuffer.Length : _buffer.Length;
+            return length - _index;
+        }
+    }
+
+    /// <summary>
+    /// Gets a span of the written data.
+    /// </summary>
+    public ReadOnlySpan<T> WrittenSpan
+    {
+        get
+        {
+            ThrowIfDisposed();
+            return _rentedBuffer is not null ? _rentedBuffer.AsSpan(0, _index) : _buffer.Slice(0, _index);
+        }
+    }
+
+    /// <summary>
+    /// Gets a memory segment of the written data.
+    /// </summary>
+    public ReadOnlyMemory<T> WrittenMemory
+    {
+        get
+        {
+            ThrowIfDisposed();
+            if (_rentedBuffer is not null)
+            {
+                return _rentedBuffer.AsMemory(0, _index);
+            }
+
+            // either we throw, or copy the Span<T> data into a heap-allocated array (via pooling), and then return Memory<T>.
+            // For demonstration, that it should not be used with Memory<T> unless done something specific, we throw here.
+            throw new InvalidOperationException("Cannot convert Span<T> to Memory<T>");
+        }
+    }
+
```

Comment:
It appears there's really no safe way to call this, and it's not part of the IBufferWriter interface anyway. The best way to ensure it doesn't get called is to delete it.

```suggestion
```